🔨(docker) fix first bootstrap can't connect to DB

igobranco opened the pull request:
Fix error `dockerflow.health.E001` with the message of `Could not connect to database`.
This only occurs from a fresh installation or on the first boot of the database.
Probably this only occurs when the dev machine has a slower IO.
Split docker compose file, per supported database
and add a dependency of the app to db, if it's healthy, directly on docker compose level.

fixes #2455

lunika commented on docker-compose.yml:
```diff
@@ -1,10 +1,4 @@
-    env_file:
-      - env.d/development/postgresql
 
   mysql:
```
You don't have to remove this service as you are redifining it in `docker-compose-mysql.yml` ?

igobranco replied:
Yes, I forgot to remove the mysql service on docker-compose.yml.

lunika commented on Makefile:
```diff
@@ -42,7 +42,7 @@ endif
 DOCKER_GID           = $(shell id -g)
 DOCKER_USER          = $(DOCKER_UID):$(DOCKER_GID)
-COMPOSE              = DOCKER_USER=$(DOCKER_USER) DB_HOST=$(DB_HOST) DB_PORT=$(DB_PORT) docker compose
+COMPOSE              = DOCKER_USER=$(DOCKER_USER) DB_HOST=$(DB_HOST) DB_PORT=$(DB_PORT) docker compose -f docker-compose.yml -f docker-compose-$(DB_HOST).yml
```
We are also using the `docker compose` command in the file `bin/_config.sh`. helper used in the `bin` directory will probably not work anymore if we don't modify the `_config.sh` file.

igobranco replied:
On `bin/_config.sh` the existing if with `DB_ENGINE` is dead code, because the file `docker/compose/development/mysql/docker-compose.yml` doesn't exist.
From what I see there isn't a proper way to run pytests on mysql locally.

I think I have fixed it.
Now is possible to run:
```bash
DB_HOST=mysql bin/pytest --reuse-db tests/apps/courses/test_templates_course_detail_rdfa.py
```
And only the mysql container is started.

Ref: https://github.com/openfun/richie/blob/7a0bf9ff20b18007529ec9f99fafbd66f266b8de/bin/_config.sh#L12

igobranco replied:
Additionally, I've added support to run the tests on sqlite locally, on a new commit on this PR.

```
DB_HOST=sqlite bin/pytest
```

lunika commented:
Thanks for your contribution. If we still use `dockerize` we can keep everything in one docker compose file. IMO it's easier to maintain in a single place instead of having to do it in multiple files.
Keeping the healthcheck on both services can be a good idea. Having the best of the two modifications you made : 
- keeping healthcheck
- everything in a single place
- not using the condition service_healthy but doing it with dockerize.

WDYT ?

igobranco commented:
@lunika 

I understand, but the `dockerize` is the reason of this problem. Next, is the explanation:
The `dockerize` just runs a check port, this means that the service has started, but it doesn't check if that service is ready to receive traffic. This is more relevant on the first startup of a database, specially on a dev system with slow IO. 
When it is replaced with docker healthcheck command, each command check if that service is ready. So I think replacing the `dockerize` with healthcheck is the way to go, but this is in my opinion.

About the split of docker compose files:
I tried to still have a single docker-compose.yml with everything, but it isn't possible to have a `depends_on` with long syntax with an if (postgresql or mysql). We need the long syntax so the app depends on a healthy database, so the only option to have multiple databases was to split the docker compose file. The env var can't be interpolated on docker compose yaml keys.
Ref: https://docs.docker.com/compose/compose-file/05-services/#depends_on

I added the health check for elasticsearch, app, redis and redis sentinel services.

The `dockerize` can be replaced with `docker compose up -d <service> --wait`. So i went ahead and replaced it.

Now, with all database services with a proper healthcheck, and using docker wait it is more robust. Se aren't going to start anything with any dependent service not ok.

jbpenrath commented:
> @lunika
> 
> I understand, but the `dockerize` is the reason of this problem. Next, is the explanation: The `dockerize` just runs a check port, this means that the service has started, but it doesn't check if that service is ready to receive traffic. This is more relevant on the first startup of a database, specially on a dev system with slow IO. When it is replaced with docker healthcheck command, each command check if that service is ready. So I think replacing the `dockerize` with healthcheck is the way to go, but this is in my opinion.
> 
> About the split of docker compose files: I tried to still have a single docker-compose.yml with everything, but it isn't possible to have a `depends_on` with long syntax with an if (postgresql or mysql). We need the long syntax so the app depends on a healthy database, so the only option to have multiple databases was to split the docker compose file. The env var can't be interpolated on docker compose yaml keys. Ref: [docs.docker.com/compose/compose-file/05-services](https://docs.docker.com/compose/compose-file/05-services/#depends_on)
> 
> I added the health check for elasticsearch, app, redis and redis sentinel services.
> 
> The `dockerize` can be replaced with `docker compose up -d <service> --wait`. So i went ahead and replaced it.
> 
> Now, with all database services with a proper healthcheck, and using docker wait it is more robust. Se aren't going to start anything with any dependent service not ok.

I found a workaround to keep a single docker compose but it's a little bit hacky and I'm not quit sure this is more elegant than create several docker-compose files. It is based on proxy service that are waiting for other service to start.

Example with postgresql:

```yaml
  postgresql-server:
    image: postgres:12
    ports:
      - "5472:5432"
    env_file:
      - env.d/development/postgresql
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready" ]
      interval: 1s
      timeout: 2s
      retries: 300
  postgresql:
    image: alpine:3.18.3
    env_file:
      - env.d/development/postgresql
    depends_on:
      postgresql-server:
        condition: service_healthy
        restart: true
  mysql-server:
    image: mysql:8.0
    env_file:
      - env.d/development/mysql
    command: --default-authentication-plugin=mysql_native_password
    healthcheck:
      test: [ "CMD", "mysqladmin" ,"ping", "-h", "localhost" ]
      interval: 1s
      timeout: 2s
      retries: 300
  mysql:
    image: alpine:3.18.3
    env_file:
      - env.d/development/mysql
    depends_on:
      mysql-server:
        condition: service_healthy
        restart: true
  app:
    build:
      context: .
      target: development
      args:
        DOCKER_USER: ${DOCKER_USER:-1000}
    image: richie:development
    env_file:
      - env.d/development/common
      - env.d/development/${DEV_ENV_FILE:-dev}
      - env.d/development/${DB_HOST:-postgresql}
    networks:
      default:
        aliases:
          - app-service
    volumes:
      - .:/app
      - ./data/static:/data/static
      - ./data/media:/data/media
      - ./data/smedia:/data/smedia
    depends_on:
      - "${DB_HOST:-postgres}"
    stdin_open: true
    tty: true
    healthcheck:
      test: ["CMD", "echo", "''", ">", "/dev/tcp/127.0.0.1/8000", "||", "exit", "1"]
      interval: 1s
      timeout: 2s
      retries: 3600
```

lunika commented:
@lunika @igobranco if the multiple docker compose files is the solution, then let's go for this.

igobranco commented:
@lunika and @jbpenrath for me this is the way to go. I rebased the code. I didn't include anything on the CHANGELOG, because this change is just relevant for internal Richie development. If you are ok with the solution, please approve it so we can merge it.
